### PR TITLE
[Merged by Bors] - port `wlog` tactic

### PR DESCRIPTION
**Pull request by thorimur**
We port the tactic `wlog`.

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Tactic/WLOG.lean [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+  `wlog h : P with H`.
+
+Typically, it is useful to use the variant `wlog h : P generalizing x y`,
+to revert certain parts of the context before creating the new goal.
+In this way, the wlog-claim `this` can be applied to `x` and `y` in different orders
+(exploiting symmetry, which is the typical use case).
+
+By default, the entire context is reverted. -/
+syntax (name := wlog) "wlog " ident " : " term
+  (" generalizing" (ppSpace colGt ident)*)? (" with " ident)? : tactic
```
Comment (digama0) [inline Mathlib/Tactic/WLOG.lean:42]:
```suggestion
syntax (name := wlog) "wlog " binderIdent " : " term
  (" generalizing" (ppSpace colGt ident)*)? (" with " binderIdent)? : tactic
```

Comment (thorimur) [reply Mathlib/Tactic/WLOG.lean:42]: Oops, didn't notice that change. :)

File: Mathlib/Tactic/WLOG.lean [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+to revert certain parts of the context before creating the new goal.
+In this way, the wlog-claim `this` can be applied to `x` and `y` in different orders
+(exploiting symmetry, which is the typical use case).
+
+By default, the entire context is reverted. -/
+syntax (name := wlog) "wlog " binderIdent " : " term
+  (" generalizing" (ppSpace colGt ident)*)? (" with " binderIdent)? : tactic
+
+elab_rules : tactic
+| `(tactic| wlog $h:ident : $P:term $[ generalizing $xs*]? $[ with $H:ident]?) => do
```
Comment (mcdoll) [inline Mathlib/Tactic/WLOG.lean:45]:
```suggestion
| `(tactic| wlog $h:ident : $P:term $[ generalizing $xs*]? $[ with $H:ident]?) =>
  withMainContext do
```

**Comment by jcommelin** [comment]
Thanks so much for porting all those tactics!

bors d+

**Pull request closed without merging**